File: src/main/java/com/joecollins/graphics/components/MultiSummaryFrameBuilder.java
package com.joecollins.graphics.components;

import com.joecollins.bindings.BindableList;
import com.joecollins.bindings.Binding;
import com.joecollins.bindings.IndexedBinding;
import java.awt.Color;
import java.util.ArrayList;
import java.util.Comparator;
import java.util.List;
import java.util.function.Function;
import java.util.stream.Collectors;
import org.apache.commons.lang3.mutable.MutableBoolean;
import org.apache.commons.lang3.tuple.Pair;

public class MultiSummaryFrameBuilder {

  private MultiSummaryFrame frame = new MultiSummaryFrame();

  public static <T> MultiSummaryFrameBuilder tooClose(
      List<T> items,
      Function<T, Binding<Boolean>> display,
      Function<T, Binding<? extends Number>> sortFunc,
      Function<T, Binding<String>> rowHeaderFunc,
      Function<T, Binding<List<Pair<Color, String>>>> rowLabelsFunc,
      int limit) {

    class Row {
      boolean display;
      double sort;
      String rowHeader;
      List<Pair<Color, String>> rowLabels;
    }

    List<Row> allRows = new ArrayList<>();
    BindableList<Row> displayedRows = new BindableList<>();

    MutableBoolean isReady = new MutableBoolean(false);
    Runnable update =
        () -> {
          if (isReady.booleanValue())
            displayedRows.setAll(
                allRows.stream()
                    .filter(r -> r.display)
                    .sorted(Comparator.comparing(r -> r.sort))
                    .limit(limit)
                    .collect(Collectors.toList()));
        };

    for (T item : items) {
      Row row = new Row();
      display
          .apply(item)
          .bind(
              d -> {
                row.display = d;
                update.run();
              });
      sortFunc
          .apply(item)
          .bind(
              s -> {
                row.sort = s.doubleValue();
                update.run();
              });
      rowHeaderFunc
          .apply(item)
          .bind(
              h -> {
                row.rowHeader = h;
                update.run();
              });
      rowLabelsFunc
          .apply(item)
          .bind(
              l -> {
                row.rowLabels = l;
                update.run();
              });
      allRows.add(row);
    }

    isReady.setTrue();
    update.run();
    MultiSummaryFrameBuilder builder = new MultiSummaryFrameBuilder();
    builder.frame.setNumRowsBinding(Binding.sizeBinding(displayedRows));
    builder.frame.setRowHeaderBinding(
        IndexedBinding.propertyBinding(displayedRows, t -> t.rowHeader));
    builder.frame.setValuesBinding(IndexedBinding.propertyBinding(displayedRows, t -> t.rowLabels));
    return builder;
  }

  public MultiSummaryFrameBuilder withHeader(Binding<String> header) {
    frame.setHeaderBinding(header);
    return this;
  }

  public MultiSummaryFrame build() {
    return frame;
  }
}

File: src/main/java/com/joecollins/graphics/components/MultiSummaryFrameBuilder.kt
package com.joecollins.graphics.components

import com.joecollins.bindings.Binding
import com.joecollins.pubsub.asOneTimePublisher
import java.awt.Color

class MultiSummaryFrameBuilder private constructor() {

    private var headerBinding: Binding<String?>? = null
    private var rowsBinding: Binding<List<MultiSummaryFrame.Row>>? = null

    fun withHeader(header: Binding<String?>): MultiSummaryFrameBuilder {
        this.headerBinding = header
        return this
    }

    fun build(): MultiSummaryFrame {
        return MultiSummaryFrame(
            headerPublisher = headerBinding?.toPublisher() ?: (null as String?).asOneTimePublisher(),
            rowsPublisher = rowsBinding?.toPublisher() ?: emptyList<MultiSummaryFrame.Row>().asOneTimePublisher()
        )
    }

    companion object {
        @JvmStatic fun <T> tooClose(
            items: List<T>,
            display: (T) -> Binding<Boolean>,
            sortFunc: (T) -> Binding<Number>,
            rowHeaderFunc: (T) -> Binding<String>,
            rowLabelsFunc: (T) -> Binding<List<Pair<Color, String>>>,
            limit: Int
        ): MultiSummaryFrameBuilder {

            class Row(val display: Boolean, val sort: Number, val row: MultiSummaryFrame.Row)

            val displayedRows: Binding<List<MultiSummaryFrame.Row>> =
                Binding.listBinding(
                    items.map {
                        val meta = display(it).merge(sortFunc(it)) { d, s -> Pair(d, s) }
                        val row = rowHeaderFunc(it).merge(rowLabelsFunc(it)) { h, v -> MultiSummaryFrame.Row(h, v) }
                        meta.merge(row) { m, r -> Row(m.first, m.second, r) }
                    }
                ).map { rows ->
                    rows.filter { it.display }
                        .sortedBy { it.sort.toDouble() }
                        .map { it.row }
                        .take(limit)
                }
            val builder = MultiSummaryFrameBuilder()
            builder.rowsBinding = displayedRows
            return builder
        }
    }
}
